### ISSUE TITLE
In `layout.tsx` that lives next to `loader.tsx`, `headers` are empty when navigated via `Link`

### ISSUE THREAD

**Issue by danieltott** [vercel/next.js#56068]
### Link to the code that reproduces this issue

https://github.com/danieltott/next-broken-linked-headers

### To Reproduce

**This does not happen when running `pnpm dev` or `pnpm build && pnpm start` locally on my mac (node v18.8.0), only on vercel**

1. Go to https://next-broken-linked-headers.vercel.app/
2. Click the "set cookie" link
3. Follow the links to see layouts that sit next to `loading.tsx` failing to read _any_ headers upon client-side navigation

### Current vs. Expected behavior

Current:
When navigating to a server component, `headers` is empty except for `"next-url: /its-broken"` (or whatever the url is).

Expected:
I can read headers/cookies in layout server components at any time.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.0.0: Fri Sep 15 14:41:34 PDT 2023; root:xnu-10002.1.13~1/RELEASE_ARM64_T8103
Binaries:
  Node: 18.8.0
  npm: 8.18.0
  Yarn: 1.22.19
  pnpm: 8.8.0
Relevant Packages:
  next: 13.5.4-canary.1
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.1.3
Next.js Config:
  output: N/A


warn  - Latest canary version not detected, detected: "13.5.4-canary.1", newest: "13.5.4-canary.0".
        Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.


Note - I don't see `13.5.4-canary.1` on the github releases page, but it is what `pnpm create next-app --example reproduction-template reproduction-app` installed.
```

### Which area(s) are affected? (Select all that apply)

App Router, Routing (next/router, next/navigation, next/link)

### Additional context

I am 99% sure this is related to the changes in #55950

This problem does _not_ happen in `13.5.3-canary.3`, but starts in `13.5.3-canary.4` and later. You can see it working properly here: https://next-broken-linked-headers-git-version-1353-canary3-danieltott.vercel.app/ with `13.5.3-canary.3` installed.

`headers` is only empty in `layout.tsx` (not `page.tsx`) and only when `layout.tsx` is next to a `loader.tsx`.

**Comment by danieltott** [comment]
I've had to pin to `13.5.3-canary.3` until this bug gets resolved.

**Comment by feliche93** [comment]
@danieltott thanks a ton for raising this. Have been debugging for several hours to find out the issue! This unblocked me 🙏

**Comment by danieltott** [comment]
@feliche93 you're welcome - I spent most of yesterday tracking and narrowing it down 😅

**Comment by christo9090** [comment]
I believe I am having the same issue but only in 13.5.3. Navigating using a next Link makes the root layout.tsx file act oddly. I am calling getServerSession for next auth. Works fine on initial load, but when navigating using a <Link>, the app shows I'm logged out. I am guessing this is related because the cookie would be missing in the layout file.

**Comment by danieltott** [comment]
Verified this issue doesn't exist in Next > 15.2.4 (at least - I didn't check 14.x).

https://next-broken-linked-headers-git-version-1524-dan-otts-projects.vercel.app/
